analyzers/src/SonarAnalyzer.CSharp/Rules/ClearTextProtocolsAreSensitive.cs

```diff
@@ -0,0 +1,167 @@
+        private const string LocalHost = "localhost";
+        private const string LocalHostIp = "127.0.0.1";
+        private const string EnableSslName = "EnableSsl";
+
+        private static readonly DiagnosticDescriptor DefaultRule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager).WithNotConfigurable();
+
+        private static readonly DiagnosticDescriptor EnableSslRule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, EnableSslMessage, RspecStrings.ResourceManager).WithNotConfigurable();
+
+        private readonly Dictionary<string, string> recommendedProtocols = new Dictionary<string, string>
+        {
+            {"telnet", "ssh"},
+            {"ftp", "sftp, scp or ftps"},
+            {"http", "https"},
+            {"clear-text SMTP", "SMTP over SSL/TLS or SMTP with STARTTLS" }
+        };
+
+        private readonly ImmutableArray<string> validServerValues = ImmutableArray.Create(LocalHost, LocalHostIp);
+
+        private readonly CSharpObjectInitializationTracker objectInitializationTracker =
+            new CSharpObjectInitializationTracker(constantValue => constantValue is bool value && value,
+                                                  ImmutableArray.Create(KnownType.System_Net_Mail_SmtpClient, KnownType.System_Net_FtpWebRequest),
+                                                  propertyName => EnableSslName == propertyName);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DefaultRule, EnableSslRule);
+
+        public ClearTextProtocolsAreSensitive() : this(AnalyzerConfiguration.Hotspot) { }
+
+        public ClearTextProtocolsAreSensitive(IAnalyzerConfiguration analyzerConfiguration) : base(analyzerConfiguration) { }
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterCompilationStartAction(ccc =>
+            {
+                if (!IsEnabled(ccc.Options))
+                {
+                    return;
+                }
+
+                context.RegisterSyntaxNodeActionInNonGenerated(VisitStringExpressions, SyntaxKind.StringLiteralExpression, SyntaxKind.InterpolatedStringExpression);
+                context.RegisterSyntaxNodeActionInNonGenerated(VisitObjectCreation, SyntaxKind.ObjectCreationExpression);
+                context.RegisterSyntaxNodeActionInNonGenerated(VisitInvocationExpression, SyntaxKind.InvocationExpression);
+                context.RegisterSyntaxNodeActionInNonGenerated(VisitAssignments, SyntaxKind.SimpleAssignmentExpression);
+            });
+
+        private void VisitObjectCreation(SyntaxNodeAnalysisContext context)
+        {
+            var objectCreation = (ObjectCreationExpressionSyntax)context.Node;
+            if (!IsServerSafe(objectCreation) && objectInitializationTracker.ShouldBeReported(objectCreation, context.SemanticModel))
```

Comment:
should this be using `if / else` ?
-  if not, could you please add a test case where both messages are being raised?
-  if it should be if/else, could you do the pattern match before the semanticModel check?